geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java

```diff
@@ -301,6 +306,106 @@ public void shouldUpdateScore_whenSettingMemberThatAlreadyExists() {
+
+  @Test
+  public void zaddIncrOptionThrowsIfIncorrectScorePair() {
+    assertThatThrownBy(() -> jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD,
+        SORTED_SET_KEY, incrOption, "1", member, "2")).hasMessageContaining(ERROR_SYNTAX);
+  }
+
+  @Test
+  public void zaddIncrOptionDoseNotAddANewMemberWithXXOption() {
+    Object result = jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, "XX",
+        incrOption, "1", member);
+    assertThat(result).isNull();
+    assertThat(jedis.zscore(SORTED_SET_KEY, member)).isNull();
+  }
+
+  private final double initial = 355.681000005;
+  private final double increment = 9554257.921450001;
+  private final double expected = initial + increment;
+
+  @Test
+  public void zaddIncrOptionIncrementsScoreForExistingMemberWithXXOption() {
+    jedis.zadd(SORTED_SET_KEY, initial, member);
+
+    Object result = jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, "XX",
+        incrOption, Coder.doubleToString(increment), member);
+    assertThat(Coder.bytesToDouble((byte[]) result)).isEqualTo(expected);
+    assertThat(jedis.zscore(SORTED_SET_KEY, member)).isEqualTo(expected);
+  }
+
+  @Test
+  public void zaddIncrOptionAddsANewMemberWithNXOption() {
+    Object result = jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, "NX",
+        incrOption, Coder.doubleToString(increment), member);
+    assertThat(Coder.bytesToDouble((byte[]) result)).isEqualTo(increment);
+
+    assertThat(jedis.zscore(SORTED_SET_KEY, member)).isEqualTo(increment);
+  }
+
+  @Test
+  public void zaddIncrOptionDoseNotIncrementScoreForAExistingMemberWithNXOption() {
+    jedis.zadd(SORTED_SET_KEY, initial, member);
+    Object result = jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, "NX",
+        incrOption, "1", member);
+    assertThat(result).isNull();
+
+    assertThat(jedis.zscore(SORTED_SET_KEY, member)).isEqualTo(initial);
+  }
+
+  @Test
+  public void zaddIncrOptionCanIncrementScoreForExistingMemberWithChangeOption() {
+    jedis.zadd(SORTED_SET_KEY, initial, member);
+    Object result = jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, "CH",
+        incrOption, Coder.doubleToString(increment), member);
+    assertThat(Coder.bytesToDouble((byte[]) result)).isEqualTo(expected);
+
+    assertThat(jedis.zscore(SORTED_SET_KEY, member)).isEqualTo(expected);
+  }
+
+  @Test
+  public void zaddIncrOptionCanIncrementScoreForExistingMember() {
+    jedis.zadd(SORTED_SET_KEY, initial, member);
+    Object result =
+        jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, incrOption,
+            Coder.doubleToString(increment), member);
+
+    assertThat(Coder.bytesToDouble((byte[]) result)).isEqualTo(expected);
+    result = jedis.zscore(SORTED_SET_KEY, member);
+    assertThat(result).isEqualTo(expected);
+  }
+
+  @Test
+  public void zaddIncrOptionCanIncrementScoreConcurrently() {
```

Comment:
might be a good idea to add a test for concurrent modification of the same key over and over again, making sure that all the increments get applied correctly. I would do that with a random increment every time, then storing that to an atomic integer.